TESTS/mbed_hal_fpga_ci_test_shield/uart/main.cpp

```diff
@@ -0,0 +1,353 @@
+    };
+    core_util_critical_section_enter();
+    serial_irq_set(&serial, RxIrq, 0);
+    core_util_critical_section_exit();
+    tester.tx_stop();
+    tester.rx_stop();
+    for (int i = 0; i < GETC_REPS; tester_buff++, i++) {
+        TEST_ASSERT_EQUAL(tester_buff, td.rx_buff[i]);
+    }
+
+    // Make sure TX IRQ was disabled during the last RX test.
+    TEST_ASSERT_EQUAL_UINT32(checksum, tester.rx_get_checksum());
+    TEST_ASSERT_EQUAL_UINT32(2 * PUTC_REPS, tester.rx_get_count());
+
+    // Cleanup
+    serial_free(&serial);
+    tester.reset();
+}
+
+void test_init_free(PinName tx, PinName rx, PinName cts = NC, PinName rts = NC)
+{
+    bool use_flow_control = (cts != NC && rts != NC) ? true : false;
+    serial_t serial;
+    serial_init(&serial, tx, rx);
+    serial_baud(&serial, 9600);
+    serial_format(&serial, 8, ParityNone, 1);
+    if (use_flow_control) {
+        serial_set_flow_control(&serial, FlowControlRTSCTS, rts, cts);
+    }
+    serial_free(&serial);
+}
+
+void test_init_free_no_fc(PinName tx, PinName rx)
+{
+    test_init_free(tx, rx);
+}
+
+template<int BAUDRATE, int DATA_BITS, SerialParity PARITY, int STOP_BITS>
+void test_common(PinName tx, PinName rx, PinName cts, PinName rts)
+{
+    uart_test_common(BAUDRATE, DATA_BITS, PARITY, STOP_BITS, tx, rx, cts, rts);
+}
+
+template<int BAUDRATE, int DATA_BITS, SerialParity PARITY, int STOP_BITS>
+void test_common_no_fc(PinName tx, PinName rx)
+{
+    uart_test_common(BAUDRATE, DATA_BITS, PARITY, STOP_BITS, tx, rx);
+}
+
+Case cases[] = {
+    // Every set of pins from every peripheral.
+    Case("init/free, FC on", all_ports<UARTPort, DefaultFormFactor, test_init_free>),
+    Case("init/free, FC off", all_ports<UARTNoFCPort, DefaultFormFactor, test_init_free_no_fc>),
+
+    // One set of pins from every peripheral.
+    Case("basic, 9600, 8N1, FC on", all_peripherals<UARTPort, DefaultFormFactor, test_common<9600, 8, ParityNone, 1> >),
+    Case("basic, 9600, 8N1, FC off", all_peripherals<UARTNoFCPort, DefaultFormFactor, test_common_no_fc<9600, 8, ParityNone, 1> >),
+
+    // One set of pins from one peripheral.
+    // baudrate
+    Case("19200, 8N1, FC on", one_peripheral<UARTPort, DefaultFormFactor, test_common<19200, 8, ParityNone, 1> >),
+    Case("19200, 8N1, FC off", one_peripheral<UARTNoFCPort, DefaultFormFactor, test_common_no_fc<19200, 8, ParityNone, 1> >),
+    Case("38400, 8N1, FC on", one_peripheral<UARTPort, DefaultFormFactor, test_common<38400, 8, ParityNone, 1> >),
+    Case("38400, 8N1, FC off", one_peripheral<UARTNoFCPort, DefaultFormFactor, test_common_no_fc<38400, 8, ParityNone, 1> >),
+    Case("115200, 8N1, FC on", one_peripheral<UARTPort, DefaultFormFactor, test_common<115200, 8, ParityNone, 1> >),
+    Case("115200, 8N1, FC off", one_peripheral<UARTNoFCPort, DefaultFormFactor, test_common_no_fc<115200, 8, ParityNone, 1> >), // DISCO_L475VG_IOT01A
+    // data bits
+    /* FIXME not supported: K64F
```

Comment:
do we need to keep this dead code here? Can it be tracked separately (removed from here, stashed somewhere else with the tracking issue for this)

Reply:
@fkjagodzinski Can you check this?

Reply:
Ok, so I checked this part -- data bits on K64F.

1. Our HAL allows setting `data_bits` and does not restrict this value in any way:
https://github.com/ARMmbed/mbed-os/blob/b55344c50b5602ee3e6c0cce748178bede3153ea/hal/serial_api.h#L137-L144
1. The HAL implementation for K64F does not really use `data_bits` param:
https://github.com/ARMmbed/mbed-os/blob/b55344c50b5602ee3e6c0cce748178bede3153ea/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/serial_api.c#L102
1. ... and this is because the `fsl_uart.c` does not have any `data_bits`-related config item:
https://github.com/ARMmbed/mbed-os/blob/b55344c50b5602ee3e6c0cce748178bede3153ea/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/drivers/fsl_uart.h#L167-L181
1. However, the `K64 Sub-Family Reference Manual` says there are two settings:
```
Programmable 8-bit or 9-bit data format
```

Initially I thought we should uncomment this block of code and expect NXP to provide a fix. Now, knowing NXP would have to update their HAL, I don't think it would be reasonable to do so. I'm not sure if Mbed OS should require support of 9-bit UART data.

@jamesbeyond, what do you think about this? Should I remove the `data_bits` part of this test?
BTW, there are two more similar `FIXME` comments regarding ST targets. Remove these as well? (don't know why they fail yet).

Reply:
I vote for yes. dead code should be removed (tracking issue to have this, with commit including this). Then it will be easy to bring this back once it is fixed

Reply:
Fixed (d2d3d5f35db56ccf5aaddf7e7ceb59f950d3908a)